payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java

```diff
@@ -398,6 +398,8 @@ public boolean apply(final PaymentTransactionModelDao input) {
                     // We can't tell where we should be in the state machine - bail (cannot be enforced by the state machine unfortunately because UNKNOWN and PLUGIN_FAILURE are both treated as EXCEPTION)
                     if (transactionToComplete.getTransactionStatus() == TransactionStatus.UNKNOWN) {
                         throw new PaymentApiException(ErrorCode.PAYMENT_INVALID_OPERATION, paymentStateContext.getTransactionType(), transactionToComplete.getTransactionStatus());
+                    } else if (transactionToComplete.getTransactionStatus() == TransactionStatus.SUCCESS ){
+                        throw new PaymentApiException(ErrorCode.PAYMENT_ACTIVE_TRANSACTION_KEY_EXISTS, paymentStateContext.getPaymentTransactionExternalKey());
```

Comment:
> It is there currently, I think you mean put it before the original Sanity checks?

Line numbers reflect the latest master (0.16.10) - please take a look at the latest code since it has changed a bit. You will need to update the PR before we can merge anyways.

> But, the `findTransactionToCompleteAndRunSanityChecks` method returned the transaction to complete. Can't invoke Janitor before it.

The `invokeJanitor` signature doesn't need `transactionToComplete` (it goes through all transactions).

> I think this is better and doable, but need more extra and duplicated code to extract the transaction from the collection if put the sanity check here.

I don't understand this - which code needs to be duplicated?

> So I think the current fixing is a better solution. How do you think?

Nope 😄 The right fix for this bug doesn't involve new logic (i.e. a new `if`) - we just need to invoke the Janitor earlier.